src/main/java/com/google/finapp/SpannerDao.java

```diff
@@ -16,133 +16,183 @@
-                      buildUpdateAccountMutation(
-                          toAccountId, accountBalances.get(toAccountId).add(amount)),
-                      buildInsertTransactionHistoryMutation(
-                          fromAccountId, amount, /* isCredit= */ true),
-                      buildInsertTransactionHistoryMutation(
-                          toAccountId, amount, /* isCredit= */ false)));
-              return null;
-            });
+
+  void moveAccountBalance(
+      ByteArray fromAccountId, ByteArray toAccountId, BigDecimal amount)
+      throws SQLException {
+    try (Connection connection = DriverManager.getConnection(this.connectionUrl)) {
+      // begin transaction
+      connection.setAutoCommit(false);
+      try (
+          PreparedStatement readStatement = connection.prepareStatement(
+              "SELECT AccountId, Balance FROM Account WHERE (AccountId = ? or AccountId = ?)"
+          );
+          PreparedStatement updateAccountStatement = connection.prepareStatement(
+              "UPDATE Account SET Balance = ? WHERE AccountId = ?"
+          );
+          // TODO: explain PENDING_COMMIT_TIMESTAMP()
+          PreparedStatement insertTransactionStatement = connection.prepareStatement(
+              "INSERT INTO TransactionHistory (AccountId, Amount, IsCredit, EventTimestamp)"
+                  + "VALUES (?, ?, ?, PENDING_COMMIT_TIMESTAMP()), "
+                  + "(?, ?, ?, PENDING_COMMIT_TIMESTAMP())")) {
+        byte[] fromAccountIdArray = fromAccountId.toByteArray();
+        byte[] toAccountIdArray = toAccountId.toByteArray();
+        BigDecimal[] accountBalances = readAccountBalances(
+            fromAccountIdArray, toAccountIdArray, readStatement);
+        updateAccount(fromAccountIdArray,
+            accountBalances[0].subtract(amount),
+            updateAccountStatement);
+        updateAccount(toAccountIdArray,
+            accountBalances[1].add(amount), updateAccountStatement);
+        updateAccountStatement.executeBatch();
+        insertTransaction(
+            fromAccountIdArray, toAccountIdArray, amount,
+            insertTransactionStatement);
+        insertTransactionStatement.executeUpdate();
+        connection.commit();
+        System.out.printf("Balance of %s moved.\n", amount.toString());
+      }
+    }
   }
 
   void getAccountMetadata(ByteArray accountId) throws SpannerException {
 
   }
 
-  private ImmutableMap<ByteArray, BigDecimal> readAccountBalances(
-      ByteArray fromAccountId, ByteArray toAccountId, TransactionContext transaction) {
-    ResultSet resultSet =
-        transaction.read(
-            "Account",
-            KeySet.newBuilder().addKey(Key.of(fromAccountId)).addKey(Key.of(toAccountId)).build(),
-            ImmutableList.of("AccountId", "Balance"));
 
-    ImmutableMap.Builder<ByteArray, BigDecimal> accountBalancesBuilder = ImmutableMap.builder();
+  private BigDecimal[] readAccountBalances(
```

Comment:
We could also use a `Pair` object, as @c24t suggested. Or `List` or something similar.

The reason the `Map` approach was abandoned was because we cannot use `Map.get` with `byte[]` objects, since equality must be done with `Arrays.equals` (so the `get` cannot find the correct key). Another way would be to convert back and forth between `ByteArray` and `byte[]` objects between functions so that we could use object equality and `Map.get` with `ByteArray`s

Reply:
I'd like to better understand the disadvantages of ByteArray, is this captured somewhere?

Reply:
I'm not sure if this specific problem is documented anywhere. There's no "disadvantages" per say, but the problem is that `ByteArray` is an object defined in `com.google.cloud`, and generic Java `PreparedStatement`s don't support them. Thus, we have to convert to `byte[]` objects to use JDBC, and `byte[]` objects have to be compared with `Arrays.equals`, while `ByteArray` objects can use the basic `equals` method. This is what I'm assuming is causing the problem for trying to use `Map.get` methods using two different `byte[]` objects that contain the same information, but are different objects in memory.

Reply:
For longevity, here is the blurb I originally had in the PR description:

> `PreparedStatement` does not support the `ByteArray` object that is in the signature for most methods. Thus, we must convert these `ByteArray` objects to `byte[]` objects. This complicates things in `moveAccountBalance` and its helper functions, where both reading and writing to the database are required. These `byte[]` objects cannot be compared with basic object equality, instead requiring `Arrays.equals`. Options include continuously switching between `byte[]` and `ByteArray` to support interacting with the database and object equality, or--as I have implemented--sticking with `byte[]` objects and changing the structure of some helper functions.